File: day-5/1.ex
lines =
  File.read!("./input.txt")
  |> String.split("\n", trim: true)

crates =
  lines
  |> Stream.take(8)
  |> Stream.map(fn line ->
    line
    |> String.to_charlist()
    |> Enum.chunk_every(4)
    |> Enum.map(&to_string/1)
    |> Enum.map(&String.trim/1)
    |> Enum.map(&String.replace(&1, "[", ""))
    |> Enum.map(&String.replace(&1, "]", ""))
  end)
  |> Enum.to_list()
  |> List.zip()
  |> Enum.map(&Tuple.to_list/1)
  |> Enum.map(&Enum.filter(&1, fn x -> x != "" end))

instructions =
  lines
  |> Stream.drop(9)
  |> Stream.map(fn line ->
    l = String.slice(line, 5..-1)
    {amount, i} = Integer.parse(l)

    [from, to] =
      i |> String.slice(6..-1) |> String.split(" to ") |> Enum.map(&String.to_integer/1)

    {amount, from, to}
  end)
  |> Enum.to_list()

Enum.reduce(instructions, crates, fn {amount, from, to}, pos ->
  from = from - 1
  to = to - 1
  {x, f} = Enum.split(Enum.at(pos, from), amount)
  t = Enum.reverse(x) ++ Enum.at(pos, to)

  pos
  |> Enum.with_index()
  |> Enum.map(fn
    {_, ^from} -> f
    {_, ^to} -> t
    {s, _} -> s
  end)
end)
|> Enum.map(&hd/1)
|> Enum.into("")
|> IO.puts()

File: day-8/1.ts
export {};

const input = await Deno.readTextFile("./input.txt");
const grid = input
  .trimEnd()
  .split("\n")
  .map((line) => line.split("").map((n) => parseInt(n)));

// console.log(grid);

// let visible = 0;

const visible = new Set();

for (let x = 0; x < grid.length; x++) {
  let h = -1;

  for (let y = 0; y < grid[x].length; y++) {
    if (grid[x][y] > h) {
      visible.add(`${y},${x}`);
      h = grid[x][y];
    }
  }

  h = -1;

  for (let y = grid[x].length - 1; y >= 0; y--) {
    if (grid[x][y] > h) {
      visible.add(`${y},${x}`);
      h = grid[x][y];
    }
  }
}

for (let y = 0; y < grid[0].length; y++) {
  let h = -1;
  for (let x = 0; x < grid.length; x++) {
    if (grid[x][y] > h) {
      visible.add(`${y},${x}`);
      h = grid[x][y];
    }
  }

  h = -1;
  for (let x = grid.length - 1; x >= 0; x--) {
    if (grid[x][y] > h) {
      visible.add(`${y},${x}`);
      h = grid[x][y];
    }
  }
}
console.log(visible.size);
